modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/pruning/PartitionPruningMetadataExtractor.java

```diff
@@ -156,57 +156,99 @@ public IgniteRel visit(IgniteTableScan rel) {
     }
 
     private static class ModifyNodeShuttle extends IgniteRelShuttle {
-        List<RexNode> projections = null;
-        List<List<RexNode>> expressions = null;
-        boolean unionRaised = false;
+        List<TargetMapping> mapping;
         IgniteRel previous = null;
         private static final Map<Class<?>, Set<Class<?>>> allowRelTransfers = new HashMap<>();
 
+        List<List<RexNode>> finalExpressions = new ArrayList<>();
+        IgniteTable table;
+
+        List<List<RexNode>> prevProjects = new ArrayList<>();
+
+        ModifyNodeShuttle(IgniteTable table) {
+            this.table = table;
+        }
+
         static {
             allowRelTransfers.put(IgniteValues.class, Set.of(IgniteProject.class, IgniteExchange.class, IgniteTrimExchange.class));
```

Comment:
transition UnionAll -> Values is missing (as well as tests I believe)

Reply:
now it\`s possible, but i can`t explain appropriate test

Reply:
`TEST("t SELECT x, x * 2, x * 3 FROM (SELECT 1 as x UNION ALL SELECT 2 as x)", TABLE_C1_C2_NULLABLE_C3)`

test case above failed with `IndexOutOfBoundsException: Index 1 out of bounds for length 1`